### PR TITLE
School user can remove placements

### PR DESCRIPTION
**Pull request by JamieCleare2525**
## Context

School Users need to ability to remove placements.

## Changes proposed in this pull request

Add Views/Controller methods/Routes/etc. to allow Schools Users the ability to remove placements.

## Guidance to review

_Review should be based on the [prototype](https://school-placements-beta-dev-73591d8bbe5f.herokuapp.com/)_
_Screenshots of Prototype (29/02/2024)_
screencapture-school-placements-beta-dev-73591d8bbe5f-herokuapp-organisations-049fc44e-6876-4884-8a77-d363173ce8d2-placements-0a252116-5410-4302-8570-a14a38c5f3d1-2024-02-29-12_51_56
screencapture-school-placements-beta-dev-73591d8bbe5f-herokuapp-organisations-049fc44e-6876-4884-8a77-d363173ce8d2-placements-0a252116-5410-4302-8570-a14a38c5f3d1-delete-2024-02-29-12_52_07
screencapture-school-placements-beta-dev-73591d8bbe5f-herokuapp-organisations-049fc44e-6876-4884-8a77-d363173ce8d2-placements-2024-02-29-12_52_16

_Note: Adding a placement is not part of this PR and therefore buttons and functionality are missing_

- Sign in as Anne
- (Assuming you have placements in her school, if not run the seed data)
- Click the "Placements" nav option
- This will display the index page of placements for that school
- Click on the link to view the first placement
- This will direct you to the show page of this placement.
- Click the "Remove placement" link
- This should redirect you the "Are you sure?" page
- Click the "Remove placement" button
- This should redirect you to the Placements index page
- AND the placement should be removed

## Link to Trello card

https://trello.com/c/SOLdIVhV/154-draft-as-a-school-user-i-can-remove-a-placement

## Screenshots

https://github.com/DFE-Digital/itt-mentor-services/assets/17162488/100d0a74-329a-4d15-8dc4-724338d91415

### CODE REVIEW

**Review by CatalinVoineag — APPROVED** [review]

File: spec/system/placements/schools/placements/remove_a_placement_spec.rb [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  let!(:placement_2) { create(:placement, school:) }
+  let!(:subject_1) { create(:subject, name: "Subject 1", subject_area: "primary") }
+  let!(:subject_2) { create(:subject, name: "Subject 2", subject_area: "primary") }
+
+  before do
+    given_i_sign_in_as_anne
+    given_a_placement_has_one_subject(placement_1, subject_1)
+    given_a_placement_has_one_subject(placement_2, subject_2)
+    when_i_visit_the_placement_show_page(placement_1)
+  end
```
Comment (ollietreend) [inline spec/system/placements/schools/placements/remove_a_placement_spec.rb:16]: I think we should be able to achieve the same thing more simply, without needing a `given_a_placement_has_one_subject` method.

```suggestion
  let!(:placement_1) { create(:placement, school:, subjects: [subject_1]) }
  let!(:placement_2) { create(:placement, school:, subjects: [subject_2]) }
  let(:school) { build(:placements_school, name: "School 1", phase: "Primary") }
  let(:subject_1) { build(:subject, name: "Subject 1", subject_area: "primary") }
  let(:subject_2) { build(:subject, name: "Subject 2", subject_area: "primary") }

  before do
    given_i_sign_in_as_anne
    when_i_visit_the_placement_show_page(placement_1)
  end
```

That is:
- `let build()` dependent objects, and only `let! create()` the primary subjects of our tests
- provide `subjects: []` to the placement, which causes the join record to be created automatically for us

**Review by ollietreend — COMMENTED** [review]
This looks really nice & tidy 🎉 

Just a few small suggestions from me.

File: app/controllers/placements/schools/placements_controller.rb [hunk 2]
```diff
@@ -9,6 +9,14 @@ def index
 
   def show; end
 
+  def remove; end
+
+  def destroy
+    placement = @school.placements.find(params.require(:id))
+    placement.destroy!
+    redirect_to placements_school_placements_path(@school), flash: { success: t(".placement_removed") }
```
[the remark this answers is not in the document]
Comment (JamieCleare2525) [reply app/controllers/placements/schools/placements_controller.rb:17]: Due to the decorator, Ruby doesn't like this.
I could remove the decorator from `set_placement`, but as `index` and `show` both need the decorator I'd prefer to keep it as is.

File: spec/system/placements/schools/placements/remove_a_placement_spec.rb (hunk 1 above)
Comment (JamieCleare2525) [reply spec/system/placements/schools/placements/remove_a_placement_spec.rb]: resolved in [ca0bad9](https://github.com/DFE-Digital/itt-mentor-services/pull/334/commits/ca0bad985a2501c00c35bb99ee622bcf1a2f3790)

File: app/models/subject.rb [hunk 3]
```diff
@@ -10,7 +10,7 @@
 #  updated_at   :datetime         not null
 #
 class Subject < ApplicationRecord
-  has_many :placement_subject_joins
+  has_many :placement_subject_joins, dependent: :destroy
```
[the remark this answers is not in the document]
Comment (JamieCleare2525) [reply app/models/subject.rb]: resolved in [ca0bad9](https://github.com/DFE-Digital/itt-mentor-services/pull/334/commits/ca0bad985a2501c00c35bb99ee622bcf1a2f3790)

File: app/controllers/placements/schools/placements_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (ollietreend) [reply app/controllers/placements/schools/placements_controller.rb:17]: Wait, what do you mean Ruby doesn't like it? 🤔

I've checked out the PR and tried this change locally, and the test suite still passes. I've also been able to do this successfully in the Rails console:

```ruby
Placement.first.decorate.destroy!
```

But maybe I'm missing something?

File: spec/system/placements/schools/placements/remove_a_placement_spec.rb [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+  def given_i_sign_in_as_anne
+    and_there_is_an_existing_user_for("Anne")
+    and_i_visit_the_sign_in_path
+    and_i_click_sign_in
+  end
+
+  def given_a_placement_has_one_subject(placement, placements_subject)
+    PlacementSubjectJoin.create!(placement:, subject: placements_subject)
+  end
```
Comment (ollietreend) [inline spec/system/placements/schools/placements/remove_a_placement_spec.rb:57]: This can be removed now.

Comment (JamieCleare2525) [reply spec/system/placements/schools/placements/remove_a_placement_spec.rb]: resolved in [646bdb8](https://github.com/DFE-Digital/itt-mentor-services/pull/334/commits/646bdb8b359916bdad8be58abb16d9054b7c50ac)

File: app/controllers/placements/schools/placements_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (JamieCleare2525) [reply app/controllers/placements/schools/placements_controller.rb]: resolved in [646bdb8](https://github.com/DFE-Digital/itt-mentor-services/pull/334/commits/646bdb8b359916bdad8be58abb16d9054b7c50ac)

**Review by ollietreend — APPROVED** [review]
Looks good to me 👍🏻

**Pull request merged**